crates/package_manager/src/add.rs

```diff
@@ -0,0 +1,96 @@
+use crate::Install;
+use derive_more::{Display, Error};
+use miette::Diagnostic;
+use pacquet_lockfile::Lockfile;
+use pacquet_npmrc::Npmrc;
+use pacquet_package_json::PackageJsonError;
+use pacquet_package_json::{DependencyGroup, PackageJson};
+use pacquet_registry::{PackageTag, PackageVersion};
+use pacquet_tarball::Cache;
+use reqwest::Client;
+
+/// This subroutine does everything `pacquet add` is supposed to do.
+#[must_use]
+pub struct Add<'a, ListDependencyGroups, DependencyGroupList>
+where
+    ListDependencyGroups: Fn() -> DependencyGroupList,
+    DependencyGroupList: IntoIterator<Item = DependencyGroup>,
+{
+    /// Shared cache that store downloaded tarballs.
+    pub tarball_cache: &'a Cache,
+    /// HTTP client to make HTTP requests.
+    pub http_client: &'a Client,
+    /// Configuration read from `.npmrc`.
+    pub config: &'static Npmrc,
+    /// Data from the `package.json` file.
+    pub package_json: &'a mut PackageJson,
+    /// Data from the `pnpm-lock.yaml` file.
+    pub lockfile: Option<&'a Lockfile>,
+    /// List of [`DependencyGroup`]s.
+    pub list_dependency_groups: ListDependencyGroups,
+    /// Name of the package to add.
+    pub package: &'a str,
+    /// Whether `--save-exact` is provided.
+    pub save_exact: bool,
```

Comment:
That's probably for future refactoring but we already have a config field, so it doesn't make sense to pass some settings separately, like this one. Actually almost all the settings from npmrc may be overwritten via CLI flags.

`save-exact` can also be set through `.npmrc`.

Reply:
Adding `save-exact` to `.npmrc` would require the ability to merge configs from CLI to `.npmrc`. This should be in another PR.